File: docs/sphinx/examples/qec/cpp/real_time_complete.cpp
/*******************************************************************************
 * Copyright (c) 2025 NVIDIA Corporation & Affiliates.                         *
 * All rights reserved.                                                        *
 *                                                                             *
 * This source code and the accompanying materials are made available under    *
 * the terms of the Apache License 2.0 which accompanies this distribution.    *
 ******************************************************************************/
// [Begin Documentation]

#include "cudaq/qec/realtime/decoding.h"
#include "cudaq/qec/realtime/decoding_config.h"

// Configure decoder before circuit execution
cudaq::qec::decoding::config::configure_decoders_from_file("decoder_config.yaml");

__qpu__ void prep0(cudaq::qec::patch logical) {
    // Your state preparation logic
    continue;
}
__qpu__ std::vector<bool> measure_stabilizers(cudaq::qec::patch logical) {
    // Your stabilizer measurement logic
    return std::vector<bool>(12, false);
}

// Quantum kernel with real-time decoding
__qpu__ void qec_circuit(int decoder_id, int num_rounds) {
    // Reset decoder state
    cudaq::qec::decoding::reset_decoder(decoder_id);

    // Allocate qubits
    cudaq::qvector data(25), ancx(12), ancz(12);
    cudaq::qec::patch logical(data, ancx, ancz);

    // Prepare logical state
    prep0(logical);

    // Syndrome extraction with real-time decoding
    for (int round = 0; round < num_rounds; ++round) {
        auto syndromes = measure_stabilizers(logical);
        cudaq::qec::decoding::enqueue_syndromes(decoder_id, syndromes);
    }

    // Get and apply corrections
    auto corrections = cudaq::qec::decoding::get_corrections(decoder_id, 1, false);
    if (corrections[0]) {
        cudaq::x(data);  // Apply correction
    }

    // Measure logical observable
    auto result = mz(data);
}

// Clean up
cudaq::qec::decoding::config::finalize_decoders();

File: docs/sphinx/examples/qec/python/real_time_complete.py
# ============================================================================ #
# Copyright (c) 2025 NVIDIA Corporation & Affiliates.                          #
# All rights reserved.                                                         #
#                                                                              #
# This source code and the accompanying materials are made available under     #
# the terms of the Apache License 2.0 which accompanies this distribution.     #
# ============================================================================ #

# [Begin Documentation]

import cudaq
import cudaq_qec as qec
from cudaq_qec import patch

# Configure decoder before circuit execution (host-side)
qec.configure_decoders_from_file("decoder_config.yaml")


# Define helper operations
@cudaq.kernel
def prep0(logical: patch):
    # Your state preparation logic
    return


@cudaq.kernel
def measure_stabilizers(logical: patch) -> list[bool]:
    # Your stabilizer measurement logic
    return [False] * 12  # 12 stabilizers for the surface code


# Quantum kernel with real-time decoding (device-side)
@cudaq.kernel
def qec_circuit(decoder_id: int, num_rounds: int):
    # Reset decoder state
    qec.reset_decoder(decoder_id)

    # Allocate qubits
    data = cudaq.qvector(25)
    ancx = cudaq.qvector(12)
    ancz = cudaq.qvector(12)
    logical = patch(data, ancx, ancz)

    # Prepare logical state
    prep0(logical)

    # Syndrome extraction with real-time decoding
    for round in range(num_rounds):
        syndromes = measure_stabilizers(logical)
        qec.enqueue_syndromes(decoder_id, syndromes)

    # Get and apply corrections
    corrections = qec.get_corrections(decoder_id, 1, False)
    if corrections[0]:
        x(data)  # Apply correction

    # Measure logical observable
    result = mz(data)


# Execute
qec_circuit(0, 10)

# Clean up (host-side)
qec.finalize_decoders()
